### PR TITLE
Add migration support to VRFV2PlusWrapper

### PR DESCRIPTION
**Pull request by kidambisrinivas**
## Motivation

- Make VRFV2PlusWrapper migrateable to the latest VRFV2PlusCoordinator

## Solution

- Add a migrate method to VRFV2PlusWrapper, which calls VRFV2PlusCoordinator's migrate method

## Testing

```
$ forge test -vvv --match-path test/v0.8/foundry/vrf/VRFV2Wrapper_Migration.t.sol
[⠰] Compiling...
No files changed, compilation skipped

Running 2 tests for test/v0.8/foundry/vrf/VRFV2Wrapper_Migration.t.sol:VRFV2PlusWrapperTest
[PASS] testMigrateWrapperLINKPayment() (gas: 546084)
Logs:
  changePrank is deprecated. Please use vm.startPrank instead.

[PASS] testMigrateWrapperNativePayment() (gas: 490588)
Logs:
  changePrank is deprecated. Please use vm.startPrank instead.

Test result: ok. 2 passed; 0 failed; 0 skipped; finished in 2.44ms
Ran 1 test suites: 2 tests passed, 0 failed, 0 skipped (2 total tests)
```

### CODE REVIEW

File: contracts/src/v0.8/dev/vrf/VRFV2PlusWrapper.sol [hunk 1]
```diff
@@ -561,4 +561,17 @@ contract VRFV2PlusWrapper is ConfirmedOwner, TypeAndVersionInterface, VRFConsume
...
+
+  function migrate(address newCoordinator) external onlyOwner {
+    IMigration(address(s_vrfCoordinator)).migrate(SUBSCRIPTION_ID, newCoordinator);
+    s_vrfCoordinator = IVRFCoordinatorV2Plus(newCoordinator);
+  }
+}
+
+interface IMigration {
+  function migrate(uint256 subId, address newCoordinator) external;
 }
```
Comment (makramkd) [inline contracts/src/v0.8/dev/vrf/VRFV2PlusWrapper.sol:577]: @jinhoonbang do you think we should move this interface definition to a separate sol file?

Comment (kidambisrinivas) [reply contracts/src/v0.8/dev/vrf/VRFV2PlusWrapper.sol]: Done. Moved to separate file `IVRFV2PlusMigration`

File: contracts/src/v0.8/dev/interfaces/IVRFV2PlusMigration.sol [hunk 2]
```diff
@@ -0,0 +1,6 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+interface IVRFV2PlusMigration {
```
Comment (makramkd) [inline contracts/src/v0.8/dev/interfaces/IVRFV2PlusMigration.sol:4]: Lets insert some doc comments here. And I just noticed that we already have a `IVRFCoordinatorV2PlusMigration`, apologies for the bad naming recommendation. Maybe just `IVRFV2PlusMigrate`?

```suggestion
/// @notice This interface is implemented by all VRF V2+ coordinators that can
/// @notice migrate subscription data to new coordinators.
interface IVRFV2PlusMigrate {
```

File: contracts/src/v0.8/dev/interfaces/IVRFV2PlusMigration.sol [hunk 3]
```diff
@@ -0,0 +1,6 @@
...
+  function migrate(uint256 subId, address newCoordinator) external;
```
Comment (makramkd) [inline contracts/src/v0.8/dev/interfaces/IVRFV2PlusMigration.sol:5]:
```suggestion
  /**
   * @notice migrate the provided subscription ID to the provided VRF coordinator
   * @notice msg.sender must be the subscription owner and newCoordinator must
   * @notice implement IVRFCoordinatorV2PlusMigration.
   * @param subId the subscription ID to migrate
   * @param newCoordinator the vrf coordinator to migrate to
  */
  function migrate(uint256 subId, address newCoordinator) external;
```

File: contracts/src/v0.8/dev/vrf/VRFV2PlusWrapper.sol [hunk 4]
```diff
@@ -561,4 +562,13 @@ contract VRFV2PlusWrapper is ConfirmedOwner, TypeAndVersionInterface, VRFConsume
...
     _;
   }
+
+  /***************************************************************************
+   * Section: Migration of VRFV2PlusWrapper to latest VRFV2PlusCoordinator
+   ***************************************************************************/
+
+  function migrate(address newCoordinator) external onlyOwner {
+    IVRFV2PlusMigration(address(s_vrfCoordinator)).migrate(SUBSCRIPTION_ID, newCoordinator);
+    s_vrfCoordinator = IVRFCoordinatorV2Plus(newCoordinator);
```
Comment (jinhoonbang) [inline contracts/src/v0.8/dev/vrf/VRFV2PlusWrapper.sol:572]: I don't think this is needed. Coordinator will call `setCoordinator()` to update `s_vrfCoordinator` - https://github.com/smartcontractkit/chainlink/blob/develop/contracts/src/v0.8/dev/vrf/VRFCoordinatorV2_5.sol#L683

File: contracts/test/v0.8/foundry/vrf/VRFV2Wrapper_Migration.t.sol [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+    bytes extraArgs,
+    address indexed sender
+  );
+
+  function testMigrateWrapperLINKPayment() public {
+    s_linkToken.transfer(address(s_consumer), DEFAULT_LINK_FUNDING);
+
+    // Get type and version.
+    assertEq(s_wrapper.typeAndVersion(), "VRFV2Wrapper 1.0.0");
+
```
Comment (jinhoonbang) [inline contracts/test/v0.8/foundry/vrf/VRFV2Wrapper_Migration.t.sol:134]: can fund the subscription beforehand to make sure migration of funds work correctly and can assert funds no longer exist on v1 coordinator

File: contracts/src/v0.8/dev/interfaces/IVRFV2PlusMigration.sol (hunk 2 above)
Comment (kidambisrinivas) [reply contracts/src/v0.8/dev/interfaces/IVRFV2PlusMigration.sol:4]: Sure, fixed this

File: contracts/src/v0.8/dev/interfaces/IVRFV2PlusMigration.sol (hunk 3 above)
Comment (kidambisrinivas) [reply contracts/src/v0.8/dev/interfaces/IVRFV2PlusMigration.sol:5]: Done

File: contracts/src/v0.8/dev/vrf/VRFV2PlusWrapper.sol (hunk 4 above)
Comment (kidambisrinivas) [reply contracts/src/v0.8/dev/vrf/VRFV2PlusWrapper.sol:572]: Ah ok, sure.

File: contracts/test/v0.8/foundry/vrf/VRFV2Wrapper_Migration.t.sol (hunk 5 above)
Comment (kidambisrinivas) [reply contracts/test/v0.8/foundry/vrf/VRFV2Wrapper_Migration.t.sol:143]: Done

**Review by jinhoonbang — APPROVED** [review]

**Pull request merged**